uc-crux-llvm/src/UCCrux/LLVM/ExprTracker.hs

```diff
@@ -0,0 +1,105 @@
+{-
+Module       : UCCrux.LLVM.ExprTracker
+Description  : Track sources of annotated symbolic expressions
+Copyright    : (c) Galois, Inc 2022
+License      : BSD3
+Maintainer   : Langston Barrett <langston@galois.com>
+Stability    : provisional
+-}
+
+{-# LANGUAGE DataKinds #-}
+{-# LANGUAGE FlexibleContexts #-}
+{-# LANGUAGE PolyKinds #-}
+
+module UCCrux.LLVM.ExprTracker
+  ( TypedSelector(..),
+    ExprTracker,
+    empty,
+    insert,
+    lookup,
+    union
+  )
+where
+
+import           Prelude hiding (lookup)
+
+import           Data.Parameterized.Ctx (Ctx)
+import           Data.Parameterized.Classes (OrdF)
+import           Data.Parameterized.Some (Some(Some))
+
+import qualified What4.Interface as What4
+
+import           Data.Map (Map)
+import qualified Data.Map as Map
+
+import           UCCrux.LLVM.Cursor (Selector, SomeInSelector(..))
+import           UCCrux.LLVM.FullType.Type (FullType, FullTypeRepr)
+
+-- | A 'Selector' together with the type of value it points to.
+--
+-- NOTE(lb): The explicit kind signature here is necessary for GHC 8.8
+-- compatibility.
+data TypedSelector m (argTypes :: Ctx (FullType m)) (ft :: FullType m)
+  = TypedSelector
+      { tsType :: FullTypeRepr m ft,
+        tsSelector :: SomeInSelector m argTypes ft
+      }
+
+-- | Track the origins of What4 expressions.
+--
+-- This map tracks where a given expression originated from, i.e. whether it was
+-- generated as part of an argument of global variable, and if so, where in said
+-- value it resides (via a 'Selector'). This is used by
+-- 'UCCrux.LLVM.Classify.classifyBadBehavior' to deduce new preconditions and
+-- attach constraints in the right place.
+--
+-- For example, if the error is an OOB write and classification looks up the
+-- pointer in this map and it appears in an argument, it will attach a
+-- precondition that that part of the argument gets allocated with more
+-- space. If it looks up the pointer and it's not in this map, it's not clear
+-- how this error could be prevented, and the error will be unclassified.
+newtype ExprTracker m sym argTypes =
+  ExprTracker
+    { getExprTracker ::
+        Map (Some (What4.SymAnnotation sym)) (Some (TypedSelector m argTypes))
```

Comment:
I just noticed the type erasure going on here with the key and value having their types erased separately. Can we document why that is important here (i.e., why is this not a `MapF`)?

Reply:
Yeah good catch, I've added a comment.